c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
+    tsk_size_t total_samples;
+    int weight;
+    tsk_tree_position_t tree_pos;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    const double sequence_length = tables->sequence_length;
+    tsk_id_t *nodes_sample_set = NULL;
+    tsk_id_t *nodes_parent = NULL;
+    tsk_id_t *nodes_time_window = NULL;
+    int *nodes_sample = NULL;
+    int *sample_count = NULL;
+    int *inside = NULL;
+    int *outside = NULL;
+    double *coalescing_pairs = NULL;
+    double *nodes_weight = NULL;
+
+    /* check inputs */
+    ret = tsk_treeseq_check_windows(self, num_windows, windows, options);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_time_windows(num_time_windows, time_windows);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = check_set_indexes(num_sample_sets, 2 * num_set_indexes, set_indexes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_check_sample_sets(
+        self, num_sample_sets, sample_set_sizes, sample_sets);
+    if (ret != 0) {
+        goto out;
+    }
+    /* TODO: set defaults? */
+
+    /* map nodes to sample sets */
+    nodes_sample_set = tsk_malloc(num_nodes * sizeof(*nodes_sample_set));
+    if (nodes_sample_set == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_sample_set_index_map(self, num_sample_sets, sample_set_sizes, sample_sets,
+        &total_samples, nodes_sample_set);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize internal state */
+    inp = (tsk_id_t) num_sample_sets;
+    col = (tsk_id_t) num_set_indexes;
+    nodes_parent = tsk_malloc(num_nodes * sizeof(*nodes_parent));
+    nodes_sample = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    sample_count = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    if (nodes_parent == NULL || nodes_sample == NULL || sample_count == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(nodes_sample, 0, num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        if (nodes_sample_set[i] != TSK_NULL) {
+            nodes_sample[i * inp + nodes_sample_set[i]] = 1;
+        }
+    }
+    tsk_memcpy(sample_count, nodes_sample,
+        num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        nodes_parent[i] = TSK_NULL;
+    }
+
+    /* map nodes to time windows */
+    nodes_time_window = tsk_malloc(num_nodes * sizeof(*nodes_time_window));
+    if (nodes_time_window == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_time_windows_index_map(
+        self, &num_time_windows, time_windows, nodes_time_window);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize outputs */
+    row = (tsk_id_t) num_time_windows;
+    inside = tsk_malloc((tsk_size_t) col * sizeof(inside));
```

Comment:
Whoops, this needs to be ``sizeof(*inside)``. This will cause obscure failures on different platforms (ask me how I know!)